File: templates/Enterprise-Template/src/typescript/generator-botbuilder-enterprise/generators/app/templates/enterprise-bot/src/dialogs/onboarding/onboardingDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    StatePropertyAccessor,
    TurnContext } from 'botbuilder';
import {
    DialogTurnResult,
    TextPrompt,
    WaterfallDialog,
    WaterfallStepContext } from 'botbuilder-dialogs';
import { BotServices } from '../../botServices';
import { EnterpriseDialog } from '../shared/enterpriseDialog';
import { OnboardingResponses } from './onboardingResponses';
import { IOnboardingState } from './onboardingState';

export class OnboardingDialog extends EnterpriseDialog {

    // Fields
    private static readonly responder: OnboardingResponses = new OnboardingResponses();
    private readonly accessor: StatePropertyAccessor<IOnboardingState>;
    private state!: IOnboardingState;

    constructor(botServices: BotServices, accessor: StatePropertyAccessor<IOnboardingState>) {
        super(botServices, OnboardingDialog.name);

        this.accessor = accessor;
        this.initialDialogId = OnboardingDialog.name;

        // tslint:disable-next-line:no-any
        const onboarding: ((sc: WaterfallStepContext<IOnboardingState>) => Promise<DialogTurnResult<any>>)[] = [
            this.askForName.bind(this),
            this.askForEmail.bind(this),
            this.askForLocation.bind(this),
            this.finishOnboardingDialog.bind(this)
        ];

        this.addDialog(new WaterfallDialog<IOnboardingState>(this.initialDialogId, onboarding));
        this.addDialog(new TextPrompt(DialogIds.namePrompt));
        this.addDialog(new TextPrompt(DialogIds.emailPrompt));
        this.addDialog(new TextPrompt(DialogIds.locationPrompt));
    }

    public async askForName(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);

        if (this.state.name) {
            return sc.next(this.state.name);
        } else {
            return sc.prompt(DialogIds.namePrompt, {
                prompt: await OnboardingDialog.responder.renderTemplate(
                    sc.context,
                    OnboardingResponses.responseIds.NamePrompt,
                    <string> sc.context.activity.locale)
            });
        }
    }

    public async askForEmail(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);
        this.state.name = sc.result;

        await OnboardingDialog.responder.replyWith(sc.context, OnboardingResponses.responseIds.HaveNameMessage, { name: this.state.name });

        return sc.prompt(DialogIds.emailPrompt, {
            prompt: await OnboardingDialog.responder.renderTemplate(
                sc.context,
                OnboardingResponses.responseIds.EmailPrompt,
                <string> sc.context.activity.locale)
        });
    }

    public async askForLocation(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);
        this.state.email = sc.result;

        await OnboardingDialog.responder.replyWith(
            sc.context, OnboardingResponses.responseIds.HaveEmailMessage, { email: this.state.email });

        return sc.prompt(DialogIds.locationPrompt, {
            prompt: await OnboardingDialog.responder.renderTemplate(
                sc.context,
                OnboardingResponses.responseIds.LocationPrompt,
                <string> sc.context.activity.locale)
        });
    }

    public async finishOnboardingDialog(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);
        this.state.location = <string> sc.result;

        await OnboardingDialog.responder.replyWith(
            sc.context,
            OnboardingResponses.responseIds.HaveLocationMessage,
            {
                name: this.state.name,
                location: this.state.location
            });

        return sc.endDialog();
    }

    private async getStateFromAccessor(context: TurnContext): Promise<IOnboardingState>  {
        const state: IOnboardingState | undefined = await this.accessor.get(context);
        if (!state) {
            const newState: IOnboardingState = {
                email: '',
                location: '',
                name: ''
            };
            await this.accessor.set(context, newState);

            return newState;
        }

        return state;
    }
}

enum DialogIds {
    namePrompt = 'namePrompt',
    emailPrompt = 'emailPrompt',
    locationPrompt =  'locationPrompt'
}

File: templates/Virtual-Assistant-Template/csharp/Template/VA/Dialogs/OnboardingDialog.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Globalization;
using System.Threading;
using System.Threading.Tasks;
using Luis;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using $safeprojectname$.Models;
using $safeprojectname$.Responses.Onboarding;
using $safeprojectname$.Services;

namespace $safeprojectname$.Dialogs
{
    public class OnboardingDialog : ComponentDialog
    {
        private static OnboardingResponses _responder = new OnboardingResponses();
        private IStatePropertyAccessor<AssistantState> _accessor;
        private AssistantState _state;
        private BotServices _services;

        public OnboardingDialog(
            BotServices botServices,
            UserState userState,
            IBotTelemetryClient telemetryClient)
            : base(nameof(OnboardingDialog))
        {
            _accessor = userState.CreateProperty<AssistantState>(nameof(AssistantState));
            InitialDialogId = nameof(OnboardingDialog);
            _services = botServices;

            var onboarding = new WaterfallStep[]
            {
                AskForName,
                FinishOnboardingDialog,
            };

            // To capture built-in waterfall dialog telemetry, set the telemetry client
            // to the new waterfall dialog and add it to the component dialog
            TelemetryClient = telemetryClient;
            AddDialog(new WaterfallDialog(InitialDialogId, onboarding) { TelemetryClient = telemetryClient });
            AddDialog(new TextPrompt(DialogIds.NamePrompt));
        }

        public async Task<DialogTurnResult> AskForName(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new AssistantState());

            if (!string.IsNullOrEmpty(_state.Name))
            {
                return await sc.NextAsync(_state.Name);
            }
            else
            {
                return await sc.PromptAsync(DialogIds.NamePrompt, new PromptOptions()
                {
                    Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.NamePrompt),
                });
            }
        }

        public async Task<DialogTurnResult> FinishOnboardingDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new AssistantState());
            var name = _state.Name = (string)sc.Result;

            var luisResult = _state.GeneralLuisResult;
            if (luisResult != null && luisResult.TopIntent().intent == GeneralLuis.Intent.ExtractName)
            {
                if (luisResult.Entities.PersonName_Any != null)
                {
                    name = _state.Name = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(luisResult.Entities.PersonName_Any[0]);
                }
                else if (luisResult.Entities.personName != null)
                {
                    name = _state.Name = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(luisResult.Entities.personName[0]);
                }
            }

            await _accessor.SetAsync(sc.Context, _state, cancellationToken);
            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveNameMessage, new { name });
            return await sc.EndDialogAsync();
        }

        private class DialogIds
        {
            public const string NamePrompt = "namePrompt";
        }
    }
}
